✨Create new extension - AMP-smartlinks

PhilWinchester opened the pull request:
Resolves https://github.com/ampproject/amphtml/issues/19470

## What it does
Create an extension that lets Narrativ offer our publishers our Linkmate service in AMP pages. The extension scans the page for any outbound links the publisher wants targeted and sends them to our API for monetization. Upon API response we use the `link-rewriter` service to map that response to the links in the initial payload. When a user clicks one of the monetized links we swap the href with the monetized link. More about Linkmate [here](http://docs.narrativ.com/en/stable/linkmate.html#linkmate).

PhilWinchester commented:
I signed it!

SHWinchester commented:
Yes!

On Fri, Jan 25, 2019 at 3:00 PM Phil Winchester <notifications@github.com>
wrote:

> I signed it!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/20494#issuecomment-457707026>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AP2wKQ0-WW2I1WUGOolB3cCPhE2ZDtgvks5vG2J2gaJpZM4aPzZp>
> .
>

honeybadgerdontcare commented on extensions/amp-smartlinks/validator-amp-smartlinks.protoascii:
```diff
@@ -0,0 +1,53 @@
+  html_format: AMP
+  tag_name: "AMP-SMARTLINKS"
+  requires_extension: "amp-smartlinks"
+  attrs: {
```
Please alphabetize by the attribute's name. E.g. `linkmate` should be before `nrtv-account-name`

honeybadgerdontcare commented on extensions/amp-smartlinks/validator-amp-smartlinks.protoascii:
```diff
@@ -0,0 +1,53 @@
+  html_format: AMP
+  tag_name: "AMP-SMARTLINKS"
+  requires_extension: "amp-smartlinks"
+  attrs: {
```
Are there limits on what the values should be for these attributes?

For instance in the markdown `linkmate` is set to `true`.

Is it just `true` or `false`? If so then it can limit the value by adding:
```
  value: "false"
  value: "true"
```

Or is it merely it's presence that determines `true` or `false`? If so it doesn't need any value and can be enforced with:
```
  value: ""
```

Consider what limits make sense for the other attributes here.

honeybadgerdontcare replied:
To follow up, if it's only true and false, then perhaps merely it's presence can be the indicator then. If attribute `exclusive-links` is present then that means true, otherwise it'll be interpreted as false. To do this then it'd be `value: ""`.

honeybadgerdontcare commented on extensions/amp-smartlinks/validator-amp-smartlinks.protoascii:
```diff
@@ -0,0 +1,53 @@
+#
+# Copyright 2018 The AMP HTML Authors. All Rights Reserved.
```
2019

honeybadgerdontcare commented on extensions/amp-smartlinks/0.1/test/validator-amp-smartlinks.out:
```diff
@@ -0,0 +1,37 @@
+|  <!--
+|    Copyright 2018 The AMP HTML Authors. All Rights Reserved.
```
2019

honeybadgerdontcare commented on extensions/amp-smartlinks/0.1/test/validator-amp-smartlinks.html:
```diff
@@ -0,0 +1,37 @@
+<!--
+	Copyright 2018 The AMP HTML Authors. All Rights Reserved.
```
2019

honeybadgerdontcare commented on extensions/amp-smartlinks/0.1/test/validator-amp-smartlinks.html:
```diff
@@ -0,0 +1,37 @@
+-->
+<!--
+	Test Description:
+	Valid amp-smartlinks tag
```
Nit: this may eventually test invalid amp-smartlinks in the same file, so perhaps instead `Tests amp-smartlinks tag` (or extension instead of tag).

honeybadgerdontcare commented on extensions/amp-smartlinks/0.1/test/validator-amp-smartlinks.html:
```diff
@@ -0,0 +1,37 @@
+	<script async src="https://cdn.ampproject.org/v0.js"></script>
+</head>
+<body>
+    <amp-smartlinks
```
We tend to prefix each test "case" with whether it is intended to be valid or invalid. When invalid we state why it's invalid.

Before this line, add `    <!-- Valid -->`

Example of invalid test:
```
    <! Invalid: missing mandatory attribute nrtv-account-name -->
    <amp-smartlinks>
        layout="nodisplay"
        linkmate="true"
        exclusive-links="false"
        link-attribute="href"
        link-selector="a">
    </amp-smartlinks>
```

PhilWinchester commented:
Thanks for the review @honeybadgerdontcare. I addressed your validator comment, please let me know if you spot anything else!

zhouyx commented:
@PhilWinchester Thank you for the PR. I am working on reviewing it, but it's a pretty big one, please allow me some time to review the change. 
And + @alabiaga who reviews the amp-skimlink change.

zhouyx commented on extensions/amp-smartlinks/0.1/amp-smartlinks.js:
```diff
@@ -0,0 +1,195 @@
+  postPageImpression_() {
+    const payload = this.buildPageImpressionPayload_();
+
+    this.xhr_.fetchJson(ENDPOINTS.PAGE_IMPRESSION_ENDPOINT, {
```
+ @lannka Ideally I would recommend them to use our `CustomEventReporterBuilder`. However I don't believe we support sending request body with analytics custom event? Do you want to prioritize that feature and get it implemented? Thanks.

lannka replied:
chat offline. we do have body payload supported in `CustomEventReporterBuilder` as a `transport` level config.

PhilWinchester replied:
Would you like me to implement this API use `CustomEventReporterBuilder`? I think I can get it working from some examples in the codebase I looked at.

zhouyx replied:
@PhilWinchester Yes please migrate to the `CustomEventReporterBuilder` [API](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-components-analytics.md#customeventreporter-api).  Let me know if there's any question. Thank you

PhilWinchester replied:
If I'm using `useBody: true` what's the best way to set the `extraUrlParams: { body: payload }`? For now I've created a helper function in `extension-analytics.js` similar to `setTransportConfig` but I'm not sure if that's the best way to go about doing this. 

```js
  postPageImpression_() {
    const builder = new CustomEventReporterBuilder(this.element);
    const payload = this.buildPageImpressionPayload_();

    builder.track('page-impression', ENDPOINTS.PAGE_IMPRESSION_ENDPOINT);
    builder.setTransportConfig(dict({
      'beacon': true,
      'xhrpost': false,
      'useBody': true,
    }));
    builder.setExtraUrlParams(payload);

    const reporter = builder.build();
    reporter.trigger('page-impression', dict());
  }
```

zhouyx replied:
@PhilWinchester That's a good point. I think adding a helper method to do that is fine for now. @lannka I know you wanted to use the `CustomEventReporter`. I do see a need to redesign the API a bit because there're many new features added to amp-analytics. If so we can create a helper function to unblock the pr for now.

zhouyx commented on extensions/amp-smartlinks/0.1/linkmate.js:
```diff
@@ -0,0 +1,179 @@
+   * @private
+   */
+  getEditInfo_() {
+    return {
```
nit: use `dict({})`, to return a JsonObject.

PhilWinchester commented on extensions/amp-smartlinks/0.1/link-rewriter/link-replacement-cache.js:
```diff
@@ -0,0 +1,101 @@
+/**
```
I wasn't sure the best way to go about doing this when starting out. I'll remove all the vendored code and switch to importing. Thanks for clearing that up!

PhilWinchester commented on extensions/amp-smartlinks/0.1/amp-smartlinks.js:
```diff
@@ -0,0 +1,195 @@
+
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    this.linkmateOptions_ = getConfigOptions(this.element);
```
I definitely can! If it would be easier I can split these out into separate variables.

PhilWinchester commented on extensions/amp-smartlinks/0.1/amp-smartlinks.js:
```diff
@@ -0,0 +1,195 @@
+      ampCors: false,
+    })
+        .then(res => res.json())
+        .then(res => {
```
Definitely will add a comment! If the return value is incorrect then the config values won't be set correctly and the API calls in `linkmate.js` will return 500. I think this is fine since anything that would cause that is a bad server side change by us and not the responsibility of `amp-smartlinks`.

alabiaga replied:
ah right. good catch. The same change should be made on the skim-links side. 

cc\ @slocka https://github.com/ampproject/amphtml/blob/master/extensions/amp-skimlinks/0.1/affiliate-link-resolver.js#L108

Thanks Yuxuan!

PhilWinchester commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -0,0 +1,91 @@
+ */
+function getNrtvAccountName_(element) {
+  const nrtvSlug = element.getAttribute('nrtv-account-name');
+  userAssert(nrtvSlug,
```
I think this is fine. If the user enters this value incorrectly then it will invalidate the API call [here](https://github.com/ampproject/amphtml/pull/20494/files#diff-b658a366becefc17d3a02d6a46bde00aR114) and none of the extensions functionality will happen. I'll update `linkmate-options.js` accordingly. Thanks for pointing this out!

PhilWinchester commented on extensions/amp-smartlinks/0.1/amp-smartlinks.js:
```diff
@@ -0,0 +1,195 @@
+
+    // We have a valid linkmate class, the config specified linkmate to run
+    // and our API is expecting linkmate to run
+    if (this.linkmate_ && this.linkmateOptions_.linkmateEnabled &&
```
Whoops I meant for this to be inside the `.then()` thanks for noticing. I put that there as a sanity check but I can't think of a scenario when `this.linkmate_` would be null. I'll update the `if`.

alabiaga commented on extensions/amp-smartlinks/0.1/constants.js:
```diff
@@ -0,0 +1,27 @@
+
+export const SMARTLINKS_REWRITER_ID = 'amp-smartlinks';
+export const BASE_API_URL = 'https://api.narrativ.com/api';
+
```
/** @enum {string} */
    export const Endpoints = {
      ....
    }

alabiaga commented on extensions/amp-smartlinks/0.1/link-rewriter/constants.js:
```diff
@@ -0,0 +1,23 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
/** @enum {string} */
export const Events = {
  ....
}

zhouyx commented:
>  My worries about importing code from that extension is that now you have this dependency on that extension which might break or change the way your extension behaves if they modify it.

@alabiaga That's a really good point. Can we reach out to the previous contributor to move the code to `src/`

PhilWinchester commented:
Thanks for the feedback and tips everyone! Any update on what is going to happen with the `link-rewriter` code?

alabiaga commented:
> Thanks for the feedback and tips everyone! Any update on what is going to happen with the `link-rewriter` code?

I reached out to @slocka and awaiting his response. Thanks.

PhilWinchester commented:
The integration test failure in travis seems to be unrelated to my changes (couldn't replicate locally). What's the best way to rerun tests?

alabiaga commented:
> The integration test failure in travis seems to be unrelated to my changes (couldn't replicate locally). What's the best way to rerun tests?

I restarted it for you and will monitor it.

PhilWinchester commented:
I'm running into issues with the one of the comments above (https://github.com/ampproject/amphtml/pull/20494#discussion_r253488460), when actually triggering the event the key isn't registered in the `this.observables_` object. If I understand the code correctly then I am falling into the `SandboxEvent` buffer and not actually triggering my event.  Ignoring the `useBody` question I mentioned above I am wondering where I am going wrong in setting up this API call. The below snippet is called after `viewer.WhenFirstVisible()`. Can someone help me understand how I am going about setting `CustomEventReporter` incorrectly? Also, any advice on `useBody` and `extraUrlParams` is greatly appreciated since our API expects the payload to be in the body object. 

```js   
/**
   * API call to indicate a page load event happened
   * @private
   */
  postPageImpression_() {
    // When using layout='nodisplay' manually trigger CustomEventReporterBuilder
    this.signals().signal(CommonSignals.LOAD_START);
    const builder = new CustomEventReporterBuilder(this.element);
    const payload = this.buildPageImpressionPayload_();

    builder.track('page-impression', ENDPOINTS.PAGE_IMPRESSION_ENDPOINT);
    builder.setTransportConfig(dict({
      'beacon': true,
      'xhrpost': false,
      'image': true,
      // 'useBody': true,
    }));
    // builder.setExtraUrlParams(payload);
    const reporter = builder.build();

    reporter.trigger('page-impression', dict({
      'data': JSON.stringify(payload),
    }));
  }
```

Separately, any update on the `link-rewriter` changes and if that is going to be a change I should be waiting for?

alabiaga commented:
@llanka or @zhouyx are best suited to answer your question about that API

> Separately, any update on the `link-rewriter` changes and if that is going to be a change I should be waiting for?

I wouldn't rely on that change to be submitted before submitting this. I would keep in contact with @slocka to see his progression on that rewrite. I would submit this with the dependency on the skimlinks link rewriter logic and follow up with @slocka to make sure that this dependency is removed eventually. With the proper tests in place, you can still be notified on whether a change on their end will require changes to your extensions but that is something you don't want to deal with.

torch2424 commented:
Hello, so I was discussing with @lannka , and we noticed that we are getting a lot more interesting in affiliate linking within AMP. And in fact, we had this proposal in the past, but got de-prioritized: #9276

Could you look through that, and see if it would work for your use case? We were thinking it may be better to make a more general solution on our end, that platforms can integrate with 😄

Thank you!

cc/ @alabiaga

zhouyx commented:
> I'm running into issues with the one of the comments above ([#20494 (comment)](https://github.com/ampproject/amphtml/pull/20494#discussion_r253488460)), when actually triggering the event the key isn't registered in the `this.observables_` object. If I understand the code correctly then I am falling into the `SandboxEvent` buffer and not actually triggering my event.

Could you please point me to the `this.observables_` object you're referring to. 
The event name is sandboxed, so that you only get the custom event of your component, not from other AMP components. The event name will be changed to `sandbox-${id}-${eventType}`, but the outgoing request will remain the same.

slocka commented:
> Hello, so I was discussing with @lannka , and we noticed that we are getting a lot more interesting in affiliate linking within AMP. And in fact, we had this proposal in the past, but got de-prioritized: #9276
> 
> Could you look through that, and see if it would work for your use case? We were thinking it may be better to make a more general solution on our end, that platforms can integrate with 😄
> 
> Thank you!
> 
> cc/ @alabiaga

@alabiaga I am happy to move the LinkRewriter related code common to amp-skimlinks and amp-smartlinks to the src folder like it initially was. I would like confirmation that the migration is still worth doing considering @torch2424 last message? Should I wait for more visibility on #9276 before making the migration?

alabiaga commented:
> @alabiaga I am happy to move the LinkRewriter related code common to amp-skimlinks and amp-smartlinks to the src folder like it initially was. I would like confirmation that the migration is still worth doing considering @torch2424 last message? Should I wait for more visibility on #9276 before making the migration?

A more general solution component that lives outside of runtime is definitely the better solution. Maybe under the guidance of @jpettitt , you can implement the component discussed in #9276 as it has already gone through design review and make it so that it is configurable to meet your needs and if not, just extend the component itself.

PhilWinchester commented:
@zhouyx I think I went down the wrong rabbit whole actually. Looking at the request headers I am sending the wrong type and the call is being rejected. What's the best way to set the `content-type: application/json` of the call in `amp-analytics`? 

Also, I am still not sure the best way to set the `body` attribute on without adding a new function to `CustomEventReporterBuilder`. Is that alright to do or am I going about this in the wrong way?

PhilWinchester commented:
@torch2424 Looking through that I think it's definitely something we could use. Right now I'm using @slocka's `link-rewriter` service which is something similar to what is discussed in that i2i. If an extension like that were to be added I think my current implementation could be refactored fairly easily to use it.

zhouyx commented:
@PhilWinchester `amp-analytics` current doesn't provide a way to set the content type. Is it possible to support `text/plain` on your server side? If not @lannka do you see the value of having amp-analytics supporting it?

jpettitt commented:
@honeybadgerdontcare & @zhouyx should this extension be mutually exclusive with `amp-skimlinks`.  Will bad things happen if there are two extensions attempting to rewrite links (seems like a guaranteed race condition) and should the validator prevent that?

PhilWinchester commented:
@jpettitt Part of the `link-rewriter` functionality is initializing rewriters in a priority queue. The user will have to specify the priority in the AMP HTML but rewriters will run in order based off of priority.
`<meta name="amp-link-rewriter-priorities" content="amp-skimlinks amp-smartlinks amp-other-affiliate">`

PhilWinchester commented:
@zhouyx I can make that change on our end. Thanks for clearing that up! Sorry to keep bothering about this but what's the best way to set `extraUrlParams`? 
```js
postPageImpression_() {
    // When using layout='nodisplay' manually trigger CustomEventReporterBuilder
    this.signals().signal(CommonSignals.LOAD_START);
    const builder = new CustomEventReporterBuilder(this.element);
    const payload = this.buildPageImpressionPayload_();

    builder.track('page-impression', ENDPOINTS.PAGE_IMPRESSION_ENDPOINT);
    builder.setTransportConfig(dict({
      'beacon': true,
      'xhrpost': false,
      'image': false,
      'useBody': true,
    }));
    builder.setExtraUrlParams(payload);
    const reporter = builder.build();

    reporter.trigger('page-impression', dict({
      'body': payload,
    }));
  }
```

Where `setExtraUrlParams` is a new function in `CustomEventReporterBuilder` that I added. Is this alright to do?

torch2424 commented:
@PhilWinchester Thanks you for letting us know!

My apologies @zhouyx didn't mean to derail the PR. 😢 

And thanks for the help everyone else! I think my question I asked after a short offline discussion with @lannka should be done elsewhere, and ask a larger question in the AMP ecosystem.

Thanks again everyone! 😄

zhouyx commented:
@PhilWinchester `setExtraUrlParams` looks good to me. Thank you for the detailed example.

PhilWinchester commented:
@zhouyx I updated the `page-impression` call to use `CustomEventReporter`. Let me know if I messed something up!

PhilWinchester commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -0,0 +1,91 @@
+  userAssert(linkmateEnabled,
+      'amp-smartlinks: linkmate is a required field');
+
+  return linkmateEnabled === 'true';
```
b/c of the `.toLowerCase` both of those vars will be type string when we check for comparison. This should let us handle booleans/string types in the config, unless I'm misunderstanding your question.

PhilWinchester commented:
@zhouyx @alabiaga Is it possible for you to take another look at the PR when you have a chance?

alabiaga commented on extensions/amp-smartlinks/0.1/linkmate.js:
```diff
@@ -0,0 +1,180 @@
+
+export class Linkmate {
+  /**
+   * @param {?../../../src/service/ampdoc-impl.AmpDoc} ampDoc
```
I believe these params should be non nullable and it looks to be the case.
    
    @param {!....amp-doc-impl.AmpDoc} ampDoc
    @param {!....xhr-impl.Xhr} xhr
    @param {!Object} linkMateOptions

Otherwise instantiating the following fields in a Linkmate object would result in errors when trying to access a field in a possibly nullable linkmateOptions and ampDoc object.

    /** @private {?boolean} */
    this.requestExclusiveLinks_ = linkmateOptions.exclusiveLinks;

    /** @private {?number} */
    this.publisherID_ = linkmateOptions.publisherID;

    /** @private {?string} */
    this.linkAttribute_ = linkmateOptions.linkAttribute;
    
    /** @private {!Document|!ShadowRoot} */
    this.rootNode_ = this.ampDoc_.getRootNode();

alabiaga commented on extensions/amp-smartlinks/0.1/linkmate.js:
```diff
@@ -0,0 +1,180 @@
+    // API response
+    let syncMappedLinks = null;
+    if (this.linkmateResponse_ && this.anchorList_ &&
+      !deepEquals(this.anchorList_, anchorList)) {
```
const anchorListChanged = this.anchorList_ && !deepEquals(this.anchorList_, anchorList);
    if (anchorListChanged) {
      if (this.linkmateResponse_) {
        syncMappedLinks = this.mapLinks_();
      } else {
         // We don't have an API response...
         const asyncMappedLinks = this.postToLinkmate_(anchorList)
         ...
      }
    } else {
       // If we didn't need to make an API
      this.anchorList_ = anchorList;
      return new TwoStepsResponse(syncMappedLinks, null);
    }

PhilWinchester replied:
@alabiaga

Thanks for the `anchorListChanged` suggestion! It makes the whole block much more readable.

I don't think I can structure the if statements the way you suggested b/c we need an initial check for when there is no API response and no `this.anchorList_`. I could add this as an `else if` on the outside `if` in your example or keep it the way I have now. Which would you prefer?

alabiaga replied:
You can just keep your changes but extract out your usage of `this.anchorList_ && !deepEquals(this.anchorList_, anchorList)` to a const for reuse. Thanks

PhilWinchester commented on extensions/amp-smartlinks/0.1/linkmate.js:
```diff
@@ -0,0 +1,180 @@
+            return {
+              anchor,
+              replacementUrl: anchor[this.linkAttribute_] === smartLink['url'] ?
+                `https://shop-links.co/${smartLink['auction_id']}/?amp=true` : null,
```
It does always exist in response object. I can implement the same catch/handle that `amp_config` call does if that would be safer.

PhilWinchester commented on extensions/amp-smartlinks/0.1/linkmate.js:
```diff
@@ -0,0 +1,180 @@
+    // If we don't have an API response or the anchor list has changed since
+    // last API call then build a new payload and post to API
+    if (!this.linkmateResponse_ ||
+      (this.anchorList_ && !deepEquals(this.anchorList_, anchorList))) {
```
From reading the `link-rewriter` implementation this function will be called anytime new links are written to the DOM. I implemented this so that this code would fit into `link-rewriter` nicely. If that isn't actually the case I can simplify this like you suggested.

zhouyx replied:
Ah! You're right, forgot about DOM change. Thank you for clarifying

PhilWinchester commented:
@zhouyx @alabiaga Thanks for the feedback and updated! Let me know if I implemented any of this incorrectly. I still have an open question around `runLinkmate` and if I should expect it to run multiple times or only once in the lifecycle of an extension?

PhilWinchester commented:
@zhouyx @alabiaga Do you mind taking a look at my newest changes and letting me know if I implemented your suggestions correctly? 

> I still have an open question around `runLinkmate` and if I should expect it to run multiple times or only once in the lifecycle of an extension?

PhilWinchester commented:
@zhouyx @alabiaga Sorry to keep bothering but I'm hoping to close this PR out as soon as reasonable. Is it possible to get another review today?

alabiaga commented on extensions/amp-smartlinks/0.1/constants.js:
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+/** {string} */
```
@const {string} here and below.

alabiaga commented on examples/amp-smartlinks.html:
```diff
@@ -0,0 +1,49 @@
+	<link rel="canonical" href="./regular-html-version.html">
+	<meta name="viewport" content="width=device-width,minimum-scale=1">
+	<meta name="amp-link-rewriter-priorities" content="amp-smartlinks">
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibil ...
```
add space(s) to format correctly here and below

PhilWinchester replied:
Thanks for the feedback. I'm not sure what you mean by add spaces. Comparing this boilerplate to other files in `examples/` it seems identical. Could you clarify what you mean by add spaces?

alabiaga replied:
...
    <meta name="amp-link-rewriter-priorities" content="amp-smartlinks">
    <style amp-boilerplate>..
    <script async custom-element="amp-smartlinks"...
    <script async src="https://cdn.ampproject.org/v0.js"></script>

style tag and the script referernce to v0 is not aligned with the rest of the markup.

PhilWinchester replied:
🤦‍♂️ completely misinterpreted your comment. Fixed!

alabiaga commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -0,0 +1,91 @@
+ */
+export function getConfigOptions(element) {
+  return {
+    nrtvSlug: getNrtvAccountName_(element),
```
all methods with trailing `_` should be marked @private.

alabiaga commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -0,0 +1,91 @@
+function getExclusiveLinksFlag_(element) {
+  const exclusiveLinks = element.getAttribute('exclusive-links');
+
+  return exclusiveLinks ? exclusiveLinks.toLowerCase() === 'true' : false;
```
simplify to

    return !!exclusiveLinks && exclusiveLinks.toLowerCase() === 'true';

zhouyx commented on extensions/amp-smartlinks/0.1/amp-smartlinks.js:
```diff
@@ -0,0 +1,195 @@
+   * @private
+   */
+  initLinkRewriter_() {
+    const options = {linkSelector: this.linkmateOptions_.linkSelector};
```
@PhilWinchester So I asked @erwinmombay offline. We don't think that closure compilier is smart enough to understand that the options is the same to the `LinkerRewriter` options params without a `typedef`. This won't be an issue right now because all fieldName is reserved today, but it will be an issue when closure compiler start to obfuscated the fieldName. 

We can either switch to using a JsonObject. But given only `linkSelector` is supported, can we instead pass a string to the constructor? If you think that makes sense I'm fine with a following PR to address the comment.

PhilWinchester replied:
@zhouyx 
Do you mean doing something like this? 
```js
    const options = /** @type {JsonObject} */
      ({'linkSelector': this.linkmateOptions_.linkSelector});
```
I also assume that this part of the extension will need to be changed to if/when the affiliate link changes are added to the AMP project and the `link-rewriter` service is deprecated. I'm not sure if that would happen sooner than the closure compiler changes. Thanks for your help!

PhilWinchester replied:
nvm the `link-rewriter` code isn't expecting a JsonObject. I think I will follow up later with these changes.

zhouyx replied:
Even with the affiliate link changes, we are going to heavily rely on the existing link-rewriter service. As said I'd imaging this class to be kept and reused.

rsimha commented:
@alabiaga Minified compilation is broken by this PR. See https://travis-ci.org/ampproject/amphtml/jobs/493915905#L610

Sadly, it's too expensive to closure-compile all extensions during PR jobs, so we only detect these breaks on `master` builds. I'm surprised `gulp check-types` didn't catch this.

PhilWinchester commented:
Sorry for breaking things :(
@alabiaga What's the best way for me to test this locally to make sure I don't break things again?

alabiaga commented:
Not a problem Phil. I am looking into this and will update you and we can merge it again.

zhouyx commented:
@rsimha @alabiaga Do you think it is due to amp-smartlinks trying to import modules from another extension. And there may be a race condition between when these two extensions gets built. 

@PhilWinchester No worries, we should document better to always test the compiled version locally. 
```
gulp dist --extensions=<your_extension_name>
gulp serve --compiled
```

PhilWinchester commented:
Hi @alabiaga & @zhouyx, 
Any update on a fix for this issue? Is there something I can do to help out? Would opening up a second PR be helpful in any way?

alabiaga commented:
@PhilWinchester I spoke with @zhouyx and she has a fix in mind that we are playing around with. Will have an update for you soon. Apologies for the delay.

alabiaga commented:
@PhilWinchester pls add the following line in https://github.com/ampproject/amphtml/blob/master/build-system/tasks/compile.js#L199

    // Needed for smart-links dep on skim-links
    'extensions/amp-skimlinks/0.1/**/*.js',

I verified that it prod version of binary will build fine with that change but pls verify yourself as well. Thanks

Thanks to @zhouyx and @erwinmombay for their help on this fix.
